analyzers/src/SonarAnalyzer.CSharp/Rules/ImplementISerializableCorrectly.cs

```diff
@@ -159,15 +159,18 @@ private static IEnumerable<SecondaryLocation> CheckConstructor(TypeDeclarationSy
         private static bool ImplementsISerializable(ITypeSymbol typeSymbol) =>
             typeSymbol != null
             && typeSymbol.IsPubliclyAccessible()
-            && typeSymbol.AllInterfaces.Any(IsOrImplementsISerializable);
+            && typeSymbol.Implements(KnownType.System_Runtime_Serialization_ISerializable)
+            && OptsInForSerialization(typeSymbol);
+
+        private static bool OptsInForSerialization(ITypeSymbol typeSymbol) =>
+            typeSymbol is INamedTypeSymbol namedTypeSymbol
+            && (namedTypeSymbol.IsSerializable() // [Serializable] is present at the types declaration
+                || namedTypeSymbol.Interfaces.Any(x => x.Is(KnownType.System_Runtime_Serialization_ISerializable)) // ISerializable is listed in the types declaration base type list
```

Comment:
If I remove this line the UTs are passing just fine, what is the difference between this one and `typeSymbol.Implements(KnownType.System_Runtime_Serialization_ISerializable)`? Can you add some tests to cover this?

Reply:
If I uncomment the line some tests fail, e.g. this expected issue from `ImplementISerializableCorrectly.CSharp10.cs` is not raised
```cs
    public struct SerializableStruct_NoAttribute : ISerializable    // Noncompliant {{Update this implementation of 'ISerializable' to conform to the recommended serialization pattern.}}
                                                                    // Secondary@-1 {{Add 'System.SerializableAttribute' attribute on 'SerializableStruct_NoAttribute' because it implements 'ISerializable'.}}
                                                                    // Secondary@-2 {{Add a 'private' constructor 'SerializableStruct_NoAttribute(SerializationInfo, StreamingContext)'.}}
    {
        private readonly NonSerializedType field = null; // Should be marked with [NonSerialized]

        public SerializableStruct_NoAttribute() { }

        public void GetObjectData(SerializationInfo info, StreamingContext context) { }
    }
```

Reply:
Yeah also the test coverage was not showing that line as non-covered if I remember correctly. Weird, most probably I just ran the ImplementISerializableCorrectly test method and not the whole batch (including C#9 & C#10). Resolving the conversation